cdap-common/src/main/java/io/cdap/cdap/common/http/AuthenticationChannelHandler.java

```diff
@@ -118,8 +125,20 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
       SecurityRequestContext.setUserIp(currentUserIp);
     }
 
+    ctx.fireChannelRead(msg);
+  }
+
+  /**
+   * If Audit logging is enabled then it sends the collection of audit events stored in {@link SecurityRequestContext}
+   * to get stored in a messaging system.
+   */
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
     try {
-      ctx.fireChannelRead(msg);
+      if (auditLoggingEnabled && msg instanceof HttpResponse) {
+        auditLogWriter.publish(SecurityRequestContext.getAuditLogQueue());
+      }
+      super.write(ctx, msg, promise);
     } finally {
       SecurityRequestContext.reset();
```

Comment:
Why did we move the reset to `write`? I suppose this would reset it earlier (on the response status writing instead of end of processing). Also are we guarantied there is a write?

Reply:
`Also are we guarantied there is a write` 
There might be cases where write might not be called. 

Adding a `try - finally` around `ctx.fireChannelRead(msg);` in the read method... that should ensure write is being called

Reply:
That's strange. Who was calling the write? Are we guarantied it's same thread? If not, we may mess up with thread locals. Also there can be special cases with streaming writes. Unless we have written guaranties from the netty docs, I think that's too much assumptions. We should 
 - Ensure we relocate data from thread local to channelhandlercontext. This is independent from threads
 - We react to both "write" and "close", picking data from the context and logging it.